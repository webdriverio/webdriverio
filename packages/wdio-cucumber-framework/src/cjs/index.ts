const {
    After,
    AfterAll,
    AfterStep,
    Before,
    BeforeAll,
    BeforeStep,

    Given,
    Then,
    When,

    defineParameterType,
    defineStep,
    setDefaultTimeout,
    setDefinitionFunctionWrapper,
    setWorldConstructor,

    DataTable,
    World,
} = require('@cucumber/cucumber')

const adapterFactory = {
    init: async function initCucumberFramework (...args: any[]) {
        const { CucumberAdapter } = await import('../index.js')
        // @ts-ignore just passing through args
        const adapter = new CucumberAdapter(...args as any)
        const instance = await adapter.init()
        return instance
    },
    After,
    AfterAll,
    AfterStep,
    Before,
    BeforeAll,
    BeforeStep,

    Given,
    Then,
    When,

    defineParameterType,
    defineStep,
    setDefaultTimeout,
    setDefinitionFunctionWrapper,
    setWorldConstructor,

    DataTable,
    World,
}

module.exports = adapterFactory
