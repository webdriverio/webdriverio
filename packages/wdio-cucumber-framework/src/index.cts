const {
    After,
    AfterAll,
    AfterStep,

    Before,
    BeforeAll,
    BeforeStep,

    Given,
    When,
    Then,

    DataTable,

    World,

    setDefaultTimeout,
    setDefinitionFunctionWrapper,
    setWorldConstructor,
    defineParameterType,
    defineStep

// eslint-disable-next-line @typescript-eslint/no-require-imports
} = require('@cucumber/cucumber')

const adapterFactory = {
    init: async function initCucumberFramework (...args: unknown[]) {
        const { CucumberAdapter } = await import('./index.js')
        // @ts-expect-error just passing through args
        const adapter = new CucumberAdapter(...args as unknown[])
        const instance = await adapter.init()
        return instance
    },
    After,
    AfterAll,
    AfterStep,

    Before,
    BeforeAll,
    BeforeStep,

    Given,
    When,
    Then,

    DataTable,

    World,

    setDefaultTimeout,
    setDefinitionFunctionWrapper,
    setWorldConstructor,
    defineParameterType,
    defineStep,

    publishCucumberReport: async function (cucumberMessageDir: string) {
        const { publishCucumberReport } = await import('./index.js')
        await publishCucumberReport(cucumberMessageDir)
    }
}

module.exports = adapterFactory
