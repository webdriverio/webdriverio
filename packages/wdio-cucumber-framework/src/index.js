import * as Cucumber from 'cucumber'
import mockery from 'mockery'
import isGlob from 'is-glob'
import glob from 'glob'
import path from 'path'

import CucumberReporter from './reporter'

import { EventEmitter } from 'events'

import { executeHooksWithArgs, testFnWrapper } from '@wdio/utils'
import { DEFAULT_OPTS } from './constants'
import { getDataFromResult, setUserHookNames } from './utils'

class CucumberAdapter {
    constructor(cid, config, specs, capabilities, reporter) {
        this.cwd = process.cwd()
        this.cid = cid
        this.specs = specs
        this.reporter = reporter
        this.capabilities = capabilities
        this.config = config
        this.cucumberOpts = Object.assign(DEFAULT_OPTS, config.cucumberOpts)
        this._hasTests = true
    }

    async init () {
        try {
            this.eventBroadcaster = new EventEmitter()

            const reporterOptions = {
                capabilities: this.capabilities,
                ignoreUndefinedDefinitions: Boolean(this.cucumberOpts.ignoreUndefinedDefinitions),
                failAmbiguousDefinitions: Boolean(this.cucumberOpts.failAmbiguousDefinitions),
                tagsInTitle: Boolean(this.cucumberOpts.tagsInTitle)
            }
            this.cucumberReporter = new CucumberReporter(this.eventBroadcaster, reporterOptions, this.cid, this.specs, this.reporter)

            const featurePathsToRun = this.config.cucumberFeaturesWithLineNumbers ? this.config.cucumberFeaturesWithLineNumbers : this.specs
            const pickleFilter = new Cucumber.PickleFilter({
                featurePaths: featurePathsToRun,
                names: this.cucumberOpts.name,
                tagExpression: this.cucumberOpts.tagExpression
            })

            this.testCases = await Cucumber.getTestCasesFromFilesystem({
                cwd: this.cwd,
                eventBroadcaster: this.eventBroadcaster,
                featurePaths: this.specs,
                order: this.cucumberOpts.order,
                pickleFilter
            })
            this._hasTests = this.testCases.length > 0
        } catch (runtimeError) {
            await executeHooksWithArgs(this.config.after, [runtimeError, this.capabilities, this.specs])
            throw runtimeError
        }

        return this
    }

    hasTests () {
        /**
         * Avoid spec filtering only if the feature is disabled explicitly
         * The feature has no impact on how framework/browser session is initialised.
         */
        return this.config.featureFlags.specFiltering === false || this._hasTests
    }

    async run () {
        let runtimeError
        let result

        try {
            this.registerRequiredModules()
            Cucumber.supportCodeLibraryBuilder.reset(this.cwd)

            /**
             * wdio hooks should be added before spec files are loaded
             */
            this.addWdioHooks(this.config)
            this.loadSpecFiles()
            this.wrapSteps(this.config)

            /**
             * we need to somehow identify is function is step or hook
             * so we wrap every user hook function
             */
            setUserHookNames(Cucumber.supportCodeLibraryBuilder.options)
            Cucumber.setDefaultTimeout(this.cucumberOpts.timeout)
            const supportCodeLibrary = Cucumber.supportCodeLibraryBuilder.finalize()

            /**
             * gets current step data: `{ uri, feature, scenario, step, sourceLocation }`
             * or `null` for some hooks.
             *
             * @return  {object|null}
             */
            this.getCurrentStep = ::this.cucumberReporter.eventListener.getCurrentStep

            const runtime = new Cucumber.Runtime({
                eventBroadcaster: this.eventBroadcaster,
                options: this.cucumberOpts,
                supportCodeLibrary,
                testCases: this.testCases
            })

            result = await runtime.start() ? 0 : 1

            /**
             * if we ignore undefined definitions we trust the reporter
             * with the fail count
             */
            if (this.cucumberOpts.ignoreUndefinedDefinitions && result) {
                result = this.cucumberReporter.failedCount
            }
        } catch (e) {
            runtimeError = e
            result = 1
        }

        await executeHooksWithArgs(this.config.after, [runtimeError || result, this.capabilities, this.specs])

        /**
         * in case the spec has a runtime error throw after the wdio hook
         */
        if (runtimeError) {
            throw runtimeError
        }

        return result
    }

    /**
     * Transpilation https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#transpilation
     * Usage: `['module']`
     * we extend it a bit with ability to init and pass configuration to modules.
     * Pass an array with path to module and its configuration instead:
     * Usage: `[['module', {}]]`
     * Or pass your own function
     * Usage: `[() => { require('ts-node').register({ files: true }) }]`
     */
    registerRequiredModules () {
        this.cucumberOpts.requireModule.map(requiredModule => {
            if (Array.isArray(requiredModule)) {
                require(requiredModule[0])(requiredModule[1])
            } else if (typeof requiredModule === 'function') {
                requiredModule()
            } else {
                require(requiredModule)
            }
        })
    }

    requiredFiles () {
        return this.cucumberOpts.require.reduce(
            (files, requiredFile) => files.concat(isGlob(requiredFile)
                ? glob.sync(requiredFile)
                : [requiredFile]
            ),
            []
        )
    }

    loadSpecFiles () {
        // we use mockery to allow people to import 'our' cucumber even though their spec files are in their folders
        // because of that we don't have to attach anything to the global object, and the current cucumber spec files
        // should just work with no changes with this framework
        mockery.enable({
            useCleanCache: false,
            warnOnReplace: false,
            warnOnUnregistered: false
        })
        mockery.registerMock('cucumber', Cucumber)
        this.requiredFiles().forEach((codePath) => {
            const filepath = path.isAbsolute(codePath)
                ? codePath
                : path.join(process.cwd(), codePath)

            // This allows rerunning a stepDefinitions file
            delete require.cache[require.resolve(filepath)]
            require(filepath)
        })
        mockery.disable()
    }

    /**
     * set `beforeScenario`, `afterScenario`, `beforeFeature`, `afterFeature`
     * @param {object} config config
     */
    addWdioHooks (config) {
        Cucumber.Before(function wdioHookBeforeScenario ({ sourceLocation, pickle }) {
            const { uri, feature } = getDataFromResult(global.result)
            return executeHooksWithArgs(config.beforeScenario, [uri, feature, pickle, sourceLocation])
        })
        Cucumber.After(function wdioHookAfterScenario ({ sourceLocation, pickle, result }) {
            const { uri, feature } = getDataFromResult(global.result)
            return executeHooksWithArgs(config.afterScenario, [uri, feature, pickle, result, sourceLocation])
        })
        Cucumber.BeforeAll(function wdioHookBeforeFeature () {
            const { uri, feature, scenarios } = getDataFromResult(global.result)
            return executeHooksWithArgs(config.beforeFeature, [uri, feature, scenarios])
        })
        Cucumber.AfterAll(function wdioHookAfterFeature () {
            const { uri, feature, scenarios } = getDataFromResult(global.result)
            return executeHooksWithArgs(config.afterFeature, [uri, feature, scenarios])
        })
    }

    /**
     * wraps step definition code with sync/async runner with a retry option
     * @param {object} config
     */
    wrapSteps (config) {
        const wrapStep = this.wrapStep
        const cid = this.cid
        const getCurrentStep = () => this.getCurrentStep()

        Cucumber.setDefinitionFunctionWrapper((fn, options = {}) => {
            /**
             * hooks defined in wdio.conf are already wrapped
             */
            if (fn.name.startsWith('wdioHook')) {
                return fn
            }

            /**
             * this flag is used to:
             * - avoid hook retry
             * - avoid wrap hooks with beforeStep and afterStep
             */
            const isStep = !fn.name.startsWith('userHook')

            const retryTest = isStep && isFinite(options.retry) ? parseInt(options.retry, 10) : 0
            return wrapStep(fn, retryTest, isStep, config, cid, getCurrentStep)
        })
    }

    /**
     * wrap step definition to enable retry ability
     * @param   {Function}  code            step definitoon
     * @param   {Number}    retryTest       amount of allowed repeats is case of a failure
     * @param   {boolean}   isStep
     * @param   {object}    config
     * @param   {string}    cid             cid
     * @param   {Function}  getCurrentStep  step definitoon
     * @return  {Function}                  wrapped step definiton for sync WebdriverIO code
     */
    wrapStep (code, retryTest = 0, isStep, config, cid, getCurrentStep) {
        return function (...args) {
            /**
             * wrap user step/hook with wdio before/after hooks
             */
            const { uri, feature } = getDataFromResult(global.result)
            const beforeFn = isStep ? config.beforeStep : config.beforeHook
            const afterFn = isStep ? config.afterStep : config.afterHook
            return testFnWrapper.call(this,
                isStep ? 'Step' : 'Hook',
                { specFn: code, specFnArgs: args },
                { beforeFn, beforeFnArgs: (context) => [uri, feature, getCurrentStep(), context] },
                { afterFn, afterFnArgs: (context) => [uri, feature, getCurrentStep(), context] },
                cid,
                retryTest)
        }
    }
}

const _CucumberAdapter = CucumberAdapter
const adapterFactory = {}

/**
 * tested by smoke tests
 */
/* istanbul ignore next */
adapterFactory.init = async function (...args) {
    const adapter = new _CucumberAdapter(...args)
    const instance = await adapter.init()
    return instance
}

export default adapterFactory
export { CucumberAdapter, adapterFactory }
