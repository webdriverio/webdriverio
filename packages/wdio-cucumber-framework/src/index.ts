import path from 'path'
import { EventEmitter } from 'events'

import mockery from 'mockery'
import isGlob from 'is-glob'
import glob from 'glob'

import * as Cucumber from '@cucumber/cucumber'
import GherkinStreams from '@cucumber/gherkin/dist/src/stream/GherkinStreams'
import EventDataCollector from '@cucumber/cucumber/lib/formatter/helpers/event_data_collector'
import { ITestCaseHookParameter } from '@cucumber/cucumber/lib/support_code_library_builder/types'
import { IRuntimeOptions } from '@cucumber/cucumber/lib/runtime'
import { Long }  from 'long'
import { IdGenerator } from '@cucumber/messages'

import { executeHooksWithArgs, testFnWrapper } from '@wdio/utils'
import type { Capabilities, Options } from '@wdio/types'

import CucumberReporter from './reporter'
import { DEFAULT_OPTS } from './constants'
import { CucumberOptions, StepDefinitionOptions, HookFunctionExtension as HookFunctionExtensionImport } from './types'
import { setUserHookNames } from './utils'

const { incrementing } = IdGenerator

class CucumberAdapter {
    private _cwd = process.cwd()
    private _newId = incrementing()
    private _cucumberOpts: Required<CucumberOptions>
    private _hasTests: boolean
    private _cucumberFeaturesWithLineNumbers: string[]
    private _eventBroadcaster: EventEmitter
    private _cucumberReporter: CucumberReporter
    private _eventDataCollector: EventDataCollector
    private _pickleFilter: Cucumber.PickleFilter

    getHookParams?: Function

    /**
     * make sure TS loads `@types/long` otherwise it won't find it in `@cucumber/messages`
     * see also https://github.com/cucumber/cucumber-js/issues/1491
     */
    never?: Long

    constructor(
        private _cid: string,
        private _config: Options.Testrunner,
        private _specs: string[],
        private _capabilities: Capabilities.RemoteCapability,
        private _reporter: EventEmitter
    ) {
        this._cucumberOpts = Object.assign({}, DEFAULT_OPTS, this._config.cucumberOpts as Required<CucumberOptions>)
        this._hasTests = true
        this._cucumberFeaturesWithLineNumbers = this._config.cucumberFeaturesWithLineNumbers || []
        this._eventBroadcaster = new EventEmitter()
        this._eventDataCollector = new EventDataCollector(this._eventBroadcaster)

        const featurePathsToRun = this._cucumberFeaturesWithLineNumbers.length > 0 ? this._cucumberFeaturesWithLineNumbers : this._specs
        this._pickleFilter = new Cucumber.PickleFilter({
            cwd: this._cwd,
            featurePaths: featurePathsToRun,
            names: this._cucumberOpts.names as string[],
            tagExpression: this._cucumberOpts.tagExpression
        })

        const reporterOptions = {
            capabilities: this._capabilities,
            ignoreUndefinedDefinitions: Boolean(this._cucumberOpts.ignoreUndefinedDefinitions),
            failAmbiguousDefinitions: Boolean(this._cucumberOpts.failAmbiguousDefinitions),
            tagsInTitle: Boolean(this._cucumberOpts.tagsInTitle),
            scenarioLevelReporter: Boolean(this._cucumberOpts.scenarioLevelReporter)
        }
        this._cucumberReporter = new CucumberReporter(this._eventBroadcaster, this._pickleFilter, reporterOptions, this._cid, this._specs, this._reporter)
    }

    async init() {
        try {
            const gherkinMessageStream = GherkinStreams.fromPaths(this._specs, {
                defaultDialect: this._cucumberOpts.featureDefaultLanguage,
                newId: this._newId
            })

            await Cucumber.parseGherkinMessageStream({
                cwd: this._cwd,
                eventBroadcaster: this._eventBroadcaster,
                gherkinMessageStream,
                eventDataCollector: this._eventDataCollector,
                order: this._cucumberOpts.order,
                pickleFilter: this._pickleFilter
            })

            this._hasTests = this._cucumberReporter.eventListener.getPickleIds(this._capabilities).length > 0
        } catch (runtimeError) {
            await executeHooksWithArgs('after', this._config.after, [runtimeError, this._capabilities, this._specs])
            throw runtimeError
        }

        /**
         * import and set options for `expect-webdriverio` assertion lib once
         * the framework was initiated so that it can detect the environment
         */
        const { setOptions } = require('expect-webdriverio')
        setOptions({
            wait: this._config.waitforTimeout, // ms to wait for expectation to succeed
            interval: this._config.waitforInterval, // interval between attempts
        })

        return this
    }

    hasTests() {
        return this._hasTests
    }

    async run() {
        let runtimeError
        let result
        try {
            this.registerRequiredModules()
            Cucumber.supportCodeLibraryBuilder.reset(this._cwd, this._newId)

            /**
             * wdio hooks should be added before spec files are loaded
             */
            this.addWdioHooks(this._config)
            this.loadSpecFiles()
            this.wrapSteps(this._config)

            /**
             * we need to somehow identify is function is step or hook
             * so we wrap every user hook function
             */
            setUserHookNames(Cucumber.supportCodeLibraryBuilder)
            Cucumber.setDefaultTimeout(this._cucumberOpts.timeout)
            const supportCodeLibrary = Cucumber.supportCodeLibraryBuilder.finalize()

            /**
             * gets current step data: `{ uri, feature, scenario, step, sourceLocation }`
             * or `null` for some hooks.
             */
            this.getHookParams = this._cucumberReporter
                .eventListener
                .getHookParams
                .bind(this._cucumberReporter.eventListener)

            const runtime = new Cucumber.Runtime({
                newId: this._newId,
                eventBroadcaster: this._eventBroadcaster,
                options: this._cucumberOpts as any as IRuntimeOptions,
                supportCodeLibrary,
                eventDataCollector: this._eventDataCollector,
                pickleIds: this._cucumberReporter!.eventListener.getPickleIds(this._capabilities)
            })

            result = await runtime.start() ? 0 : 1

            /**
             * if we ignore undefined definitions we trust the reporter
             * with the fail count
             */
            if (this._cucumberOpts.ignoreUndefinedDefinitions && result) {
                result = this._cucumberReporter.failedCount
            }
        } catch (e) {
            runtimeError = e
            result = 1
        }

        await executeHooksWithArgs('after', this._config.after, [runtimeError || result, this._capabilities, this._specs])

        /**
         * in case the spec has a runtime error throw after the wdio hook
         */
        if (runtimeError) {
            throw runtimeError
        }

        return result
    }

    /**
     * Transpilation https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#transpilation
     * Usage: `['module']`
     * we extend it a bit with ability to init and pass configuration to modules.
     * Pass an array with path to module and its configuration instead:
     * Usage: `[['module', {}]]`
     * Or pass your own function
     * Usage: `[() => { require('@babel/register')({ ignore: [] }) }]`
     */
    registerRequiredModules() {
        this._cucumberOpts.requireModule.map(requiredModule => {
            if (Array.isArray(requiredModule)) {
                require(requiredModule[0])(requiredModule[1])
            } else if (typeof requiredModule === 'function') {
                (requiredModule as Function)()
            } else {
                require(requiredModule)
            }
        })
    }

    requiredFiles() {
        return this._cucumberOpts.require.reduce(
            (files: string[], requiredFile: string) => files.concat(isGlob(requiredFile)
                ? glob.sync(requiredFile)
                : [requiredFile]
            ),
            []
        )
    }

    loadSpecFiles() {
        // we use mockery to allow people to import 'our' cucumber even though their spec files are in their folders
        // because of that we don't have to attach anything to the global object, and the current cucumber spec files
        // should just work with no changes with this framework
        mockery.enable({
            useCleanCache: false,
            warnOnReplace: false,
            warnOnUnregistered: false
        })
        mockery.registerMock('@cucumber/cucumber', Cucumber)
        this.requiredFiles().forEach((codePath) => {
            const filepath = path.isAbsolute(codePath)
                ? codePath
                : path.join(process.cwd(), codePath)

            // This allows rerunning a stepDefinitions file
            delete require.cache[require.resolve(filepath)]
            require(filepath)
        })
        mockery.disable()
    }

    /**
     * set `beforeScenario`, `afterScenario`, `beforeFeature`, `afterFeature`
     * @param {object} config config
     */
    addWdioHooks (config: Options.Testrunner) {
        const eventListener = this._cucumberReporter?.eventListener
        Cucumber.BeforeAll(async function wdioHookBeforeFeature() {
            const params = eventListener?.getHookParams()
            await executeHooksWithArgs(
                'beforeFeature',
                config.beforeFeature,
                [params?.uri, params?.feature]
            )
        })
        Cucumber.AfterAll(async function wdioHookAfterFeature() {
            const params = eventListener?.getHookParams()
            await executeHooksWithArgs(
                'afterFeature',
                config.afterFeature,
                [params?.uri, params?.feature]
            )
        })
        Cucumber.Before(async function wdioHookBeforeScenario (world: ITestCaseHookParameter) {
            await executeHooksWithArgs(
                'beforeScenario',
                config.beforeScenario,
                [world]
            )
        })
        Cucumber.After(async function wdioHookAfterScenario (world: ITestCaseHookParameter) {
            await executeHooksWithArgs(
                'afterScenario',
                config.afterScenario,
                [world]
            )
        })
    }

    /**
     * wraps step definition code with sync/async runner with a retry option
     * @param {object} config
     */
    wrapSteps (config: Options.Testrunner) {
        const wrapStep = this.wrapStep
        const cid = this._cid
        const getHookParams = () => this.getHookParams && this.getHookParams()

        Cucumber.setDefinitionFunctionWrapper((fn: Function, options: StepDefinitionOptions = { retry: 0 }) => {
            /**
             * hooks defined in wdio.conf are already wrapped
             */
            if (fn.name.startsWith('wdioHook')) {
                return fn
            }

            /**
             * this flag is used to:
             * - avoid hook retry
             * - avoid wrap hooks with beforeStep and afterStep
             */
            const isStep = !fn.name.startsWith('userHook')

            return wrapStep(fn, isStep, config, cid, options, getHookParams)
        })
    }

    /**
     * wrap step definition to enable retry ability
     * @param   {Function}  code            step definition
     * @param   {Number}    retryTest       amount of allowed repeats is case of a failure
     * @param   {boolean}   isStep
     * @param   {object}    config
     * @param   {string}    cid             cid
     * @param   {Function}  getHookParams  step definition
     * @return  {Function}                  wrapped step definition for sync WebdriverIO code
     */
    wrapStep(
        code: Function,
        isStep: boolean,
        config: Options.Testrunner,
        cid: string,
        options: StepDefinitionOptions,
        getHookParams: Function
    ) {
        return function (this: Cucumber.World, ...args: any[]) {
            const hookParams = getHookParams()
            const retryTest = isStep && isFinite(options.retry) ? options.retry : 0

            /**
             * wrap user step/hook with wdio before/after hooks
             */
            const beforeFn = isStep ? config.beforeStep : config.beforeHook
            const afterFn = isStep ? config.afterStep : config.afterHook
            return testFnWrapper.call(this,
                isStep ? 'Step' : 'Hook',
                { specFn: code, specFnArgs: args },
                { beforeFn: beforeFn as Function[], beforeFnArgs: (context) => [hookParams?.step, context] },
                { afterFn: afterFn as Function[], afterFnArgs: (context) => [hookParams?.step, context] },
                cid,
                retryTest)
        }
    }
}

const _CucumberAdapter = CucumberAdapter
const adapterFactory: { init?: Function } = {}

/**
 * tested by smoke tests
 */
/* istanbul ignore next */
adapterFactory.init = async function (...args: any[]) {
    // @ts-ignore just passing through args
    const adapter = new _CucumberAdapter(...args as any)
    const instance = await adapter.init()
    return instance
}

export default adapterFactory
export { CucumberAdapter, adapterFactory }

declare global {
    namespace WebdriverIO {
        interface CucumberOpts extends CucumberOptions {}
        interface HookFunctionExtension extends HookFunctionExtensionImport {}
    }
}
