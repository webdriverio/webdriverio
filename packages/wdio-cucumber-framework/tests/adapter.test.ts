import { setOptions } from 'expect-webdriverio'
import { executeHooksWithArgs, testFnWrapper } from '@wdio/utils'
import * as Cucumber from '@cucumber/cucumber'
import mockery from 'mockery'

import CucumberAdapter from '../src'
import { setUserHookNames } from '../src/utils'

jest.mock('../src/reporter', () => class CucumberReporter {
    eventListener = {
        getPickleIds: jest.fn().mockReturnValue(['8']),
        getHookParams: jest.fn().mockReturnValue({ uri: 'uri', feature: 'feature' })
    }
})

jest.mock('../src/utils', () => ({
    setUserHookNames: jest.fn()
}))

jest.mock('@cucumber/gherkin/dist/src/stream/GherkinStreams', () => ({
    fromPaths: jest.fn().mockReturnValue('GherkinStreams.fromPaths')
}))

declare global {
    var MODULE_A_WAS_LOADED: boolean
    var MODULE_B_WAS_LOADED_WITH: any
    var MODULE_INLINE_WAS_LOADED: boolean
    var MODULE_C_WAS_LOADED: boolean
}

describe('CucumberAdapter', () => {
    beforeEach(() => {
        (Cucumber.PickleFilter as jest.Mock).mockClear()
        ;(executeHooksWithArgs as jest.Mock).mockClear()
        ;(Cucumber.Runtime as jest.Mock).mockClear()
        ;(Cucumber.setDefinitionFunctionWrapper as jest.Mock).mockClear()
        ;(mockery.enable as jest.Mock).mockClear()
        ;(mockery.registerMock as jest.Mock).mockClear()
        ;(mockery.disable as jest.Mock).mockClear()
        ;(Cucumber.BeforeAll as jest.Mock).mockClear()
        ;(Cucumber.AfterAll as jest.Mock).mockClear()
        ;(Cucumber.Before as jest.Mock).mockClear()
        ;(Cucumber.After as jest.Mock).mockClear()
    })

    it('can be initated with tests', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            waitforTimeout: 1,
            waitforInterval: 2
        }, ['/foo/bar'], {}, {})

        expect(executeHooksWithArgs).toBeCalledTimes(0)
        expect(setOptions).toBeCalledWith({
            wait: 1,
            interval: 2
        })
        expect(adapter.hasTests()).toBe(true)
        expect(Cucumber.PickleFilter).toBeCalledWith({
            cwd: expect.any(String),
            featurePaths: ['/foo/bar'],
            names: [],
            tagExpression: ''
        })
    })

    it('should trigger after hook if initiation fails', async () => {
        (Cucumber.parseGherkinMessageStream as jest.Mock)
            .mockRejectedValueOnce(new Error('boom'))

        const err = await CucumberAdapter.init('0-0', {
            cucumberFeaturesWithLineNumbers: ['/bar/foo', '/foo/bar']
        }, ['/foo/bar'], {}, {})
            .catch((err) => err)
        expect(err.message).toBe('boom')
        expect(executeHooksWithArgs).toBeCalledTimes(1)
        expect(Cucumber.PickleFilter).toBeCalledWith({
            cwd: expect.any(String),
            featurePaths: ['/bar/foo', '/foo/bar'],
            names: [],
            tagExpression: ''
        })
    })

    it('can run without errors', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        adapter.registerRequiredModules = jest.fn()
        adapter.addWdioHooks = jest.fn()
        adapter.loadSpecFiles = jest.fn()
        adapter.wrapSteps = jest.fn()

        const result = await adapter.run()
        expect(result).toBe(0)

        expect(adapter.registerRequiredModules).toBeCalledTimes(1)
        expect(adapter.addWdioHooks).toBeCalledTimes(1)
        expect(adapter.loadSpecFiles).toBeCalledTimes(1)
        expect(adapter.wrapSteps).toBeCalledTimes(1)
        expect(setUserHookNames).toBeCalledTimes(1)
    })

    it('can run with failing result', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            cucumberOpts: { shouldFail: 123 }
        }, ['/foo/bar'], {}, {})
        const result = await adapter.run()
        expect(result).toBe(1)
        expect(executeHooksWithArgs).toBeCalledTimes(1)
    })

    it('can take cucumber reporte failure count', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            cucumberOpts: {
                shouldFail: 123,
                ignoreUndefinedDefinitions: true
            }
        }, ['/foo/bar'], {}, {})
        adapter._cucumberReporter.failedCount = 123
        const result = await adapter.run()
        expect(result).toBe(123)
        expect(executeHooksWithArgs).toBeCalledTimes(1)
    })

    it('should throw if there was a runtime error', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            cucumberOpts: { shouldThrow: 'some error' }
        }, ['/foo/bar'], {}, {})
        const err = await adapter.run().catch(err => err)
        expect(err.message).toBe('some error')
    })

    it('registerRequiredModules', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            cucumberOpts: {
                requireModule: [
                    'moduleA',
                    ['moduleB', { foo: 'bar' }],
                    () => {
                        // @ts-ignore
                        global.MODULE_INLINE_WAS_LOADED = true
                    }
                ]
            }
        }, ['/foo/bar'], {}, {})
        expect(global.MODULE_A_WAS_LOADED).toBe(undefined)
        expect(global.MODULE_A_WAS_LOADED).toBe(undefined)
        expect(global.MODULE_INLINE_WAS_LOADED).toBe(undefined)
        adapter.registerRequiredModules()
        expect(global.MODULE_A_WAS_LOADED).toBe(true)
        expect(global.MODULE_B_WAS_LOADED_WITH).toEqual({ foo: 'bar' })
        expect(global.MODULE_INLINE_WAS_LOADED).toBe(true)
    })

    it('requiredFiles', async () => {
        const adapter = await CucumberAdapter.init('0-0', {
            cucumberOpts: {
                require: [
                    __dirname + '/' + __filename,
                    __dirname + '/__mocks__/module*.ts'
                ]
            }
        }, ['/foo/bar'], {}, {})
        expect(adapter.requiredFiles()).toHaveLength(4)
    })

    it('loadSpecFiles', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        adapter.requiredFiles = jest.fn().mockReturnValue([__dirname + '/__mocks__/moduleC.ts'])

        expect(global.MODULE_C_WAS_LOADED).toBe(undefined)
        adapter.loadSpecFiles()
        expect(global.MODULE_C_WAS_LOADED).toBe(true)

        expect(mockery.enable).toBeCalledTimes(1)
        expect(mockery.registerMock).toBeCalledWith('@cucumber/cucumber', expect.any(Object))
        expect(mockery.disable).toBeCalledTimes(1)
    })

    it('addWdioHooks', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        adapter.addWdioHooks({
            beforeFeature: 'beforeFeature',
            afterFeature: 'afterFeature',
            beforeScenario: 'beforeScenario',
            afterScenario: 'afterScenario'
        })
        expect(Cucumber.BeforeAll).toBeCalledTimes(1)
        expect(Cucumber.AfterAll).toBeCalledTimes(1)
        expect(Cucumber.Before).toBeCalledTimes(1)
        expect(Cucumber.After).toBeCalledTimes(1)
        expect(executeHooksWithArgs).toBeCalledTimes(0)

        ;(Cucumber.BeforeAll as jest.Mock).mock.calls[0][0]()
        expect(executeHooksWithArgs)
            .toBeCalledWith('beforeFeature', 'beforeFeature', ['uri', 'feature'])
        ;(Cucumber.AfterAll as jest.Mock).mock.calls[0][0]()
        expect(executeHooksWithArgs)
            .toBeCalledWith('afterFeature', 'afterFeature', ['uri', 'feature'])
        ;(Cucumber.Before as jest.Mock).mock.calls[0][0]('world')
        expect(executeHooksWithArgs)
            .toBeCalledWith('beforeScenario', 'beforeScenario', ['world'])
        ;(Cucumber.After as jest.Mock).mock.calls[0][0]('world')
        expect(executeHooksWithArgs)
            .toBeCalledWith('afterScenario', 'afterScenario', ['world'])
    })

    it('wrapSteps', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        adapter.getHookParams = 'getHookParams'
        adapter.wrapStep = jest.fn()

        expect(adapter.wrapStep).toBeCalledTimes(0)
        adapter.wrapSteps()
        expect(Cucumber.setDefinitionFunctionWrapper).toBeCalledTimes(1)
        ;(Cucumber.setDefinitionFunctionWrapper as jest.Mock).mock.calls[0][0](jest.fn())
        expect(adapter.wrapStep).toBeCalledWith(
            expect.any(Function),
            true,
            undefined,
            '0-0',
            { retry: 0 },
            expect.any(Function)
        )
    })

    it('wrapSteps does not wrap wdio hooks', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        adapter.getHookParams = 'getHookParams'
        adapter.wrapStep = jest.fn()

        expect(adapter.wrapStep).toBeCalledTimes(0)
        adapter.wrapSteps()
        function wdioHookFn () { return 'foobar' }
        expect(Cucumber.setDefinitionFunctionWrapper).toBeCalledTimes(1)
        expect(
            (Cucumber.setDefinitionFunctionWrapper as jest.Mock).mock.calls[0][0](wdioHookFn)()
        ).toBe('foobar')
    })

    it('wrapStep', async () => {
        const adapter = await CucumberAdapter.init('0-0', {}, ['/foo/bar'], {}, {})
        const wrappedStep = adapter.wrapStep('code', true, {}, '0-2', {}, () => 'hookParams')
        expect(testFnWrapper).toBeCalledTimes(0)
        wrappedStep('someWorld', 1, 2, 3)
        expect((testFnWrapper as jest.Mock).mock.calls).toMatchSnapshot()
    })
})
