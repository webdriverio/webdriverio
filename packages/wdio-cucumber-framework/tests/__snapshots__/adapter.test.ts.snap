// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CucumberAdapter wrapStep 1`] = `
Array [
  Array [
    "Step",
    Object {
      "specFn": "code",
      "specFnArgs": Array [
        "someWorld",
        1,
        2,
        3,
      ],
    },
    Object {
      "beforeFn": undefined,
      "beforeFnArgs": [Function],
    },
    Object {
      "afterFn": undefined,
      "afterFnArgs": [Function],
    },
    "0-2",
    0,
  ],
]
`;
